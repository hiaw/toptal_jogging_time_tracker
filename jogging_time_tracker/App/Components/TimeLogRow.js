import React from 'react'
import { StyleSheet, Text, View } from 'react-native'
import moment from 'moment'
import _ from 'lodash'

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 20,
    paddingVertical: 5,
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: 'black',
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  firstRow: {
    justifyContent: 'space-between',
  },
  secondRow: {
    justifyContent: 'flex-start',
  },
  welcome: {
    fontSize: 20,
  },
  duration: {
    fontSize: 20,
    marginLeft: 50,
  },
})

const getDistanceText = distance => {
  const kms = _.round(distance / 1000, 3)
  const meters = _.round(distance, 2)
  let distanceText = `${meters} m`
  if (distance > 1000) distanceText = `${kms} km`
  return distanceText
}

const MINUTE_PER_HOUR = 3600

const padNumber = number => {
  let padded = number + ''
  if (number < 10) {
    padded = '0' + padded
  }
  return padded
}

const getDurationText = duration => {
  const hours = Math.floor(duration / MINUTE_PER_HOUR)
  const minutes = Math.floor(duration % MINUTE_PER_HOUR / 60)
  const seconds = _.round(duration % 60, 1)
  const durationText = `${hours}:${padNumber(minutes)}:${padNumber(seconds)}`
  return durationText
}

const getSpeedText = (distance, duration) => {
  const kms = distance / 1000
  const hours = duration / MINUTE_PER_HOUR
  const speed = _.round(kms / hours, 2)
  const speedText = `${speed} km/h`
  return speedText
}

const TimeLogList = props => {
  const { date, distance, duration } = props
  const dateText = moment(date).format('DD/MM/YYYY')
  const distanceText = getDistanceText(distance)
  const durationText = getDurationText(duration)
  const speedText = getSpeedText(distance, duration)
  return (
    <View style={styles.container}>
      <View style={[styles.row, styles.firstRow]}>
        <Text style={styles.welcome}>
          {dateText}
        </Text>
        <Text style={styles.welcome}>
          {speedText}
        </Text>
      </View>
      <View style={[styles.row, styles.secondRow]}>
        <Text style={styles.welcome}>
          {distanceText}
        </Text>
        <Text style={styles.duration}>
          {durationText}
        </Text>
      </View>
    </View>
  )
}

export default TimeLogList
