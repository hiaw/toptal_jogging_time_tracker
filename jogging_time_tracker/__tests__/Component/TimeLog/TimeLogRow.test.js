import 'react-native'
import React from 'react'

import TimeLogRow from '../../../App/Components/TimeRow/TimeLogRow.js'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

const noop = () => {}

it('renders correctly', () => {
  const tree = renderer
    .create(
      <TimeLogRow
        date="2017-06-23T04:51:34.022Z"
        distance={1034.3}
        duration={34.3}
        onPress={noop}
      />,
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
