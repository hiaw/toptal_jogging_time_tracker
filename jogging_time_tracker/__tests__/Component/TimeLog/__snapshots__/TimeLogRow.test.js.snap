// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Object {
        "borderBottomColor": "black",
        "borderBottomWidth": 0.5,
        "paddingHorizontal": 20,
        "paddingVertical": 5,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          Object {
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 20,
          }
        }
      >
        23/06/2017
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 20,
          }
        }
      >
        3.6 km/h
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          Object {
            "justifyContent": "flex-start",
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 20,
          }
        }
      >
        34.3 m
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 20,
            "marginLeft": 50,
          }
        }
      >
        0:00:34.3
      </Text>
    </View>
  </View>
</View>
`;
